src/server/game/Chat/Hyperlinks.cpp

```diff
@@ -0,0 +1,404 @@
+// Validates a single hyperlink
+HyperlinkInfo Trinity::Hyperlinks::ParseHyperlink(char const* pos)
+{
+    //color tag
+    if (*(pos++) != '|' || *(pos++) != 'c')
+        return nullptr;
+    uint32 color = 0;
+    for (uint8 i = 0; i < 8; ++i)
+        if (uint8 hex = toHex(*(pos++)))
+            color = (color << 4) | (hex & 0xf);
+    // link data start tag
+    if (*(pos++) != '|' || *(pos++) != 'H')
+        return nullptr;
+    // link tag, find next : or |
+    char const* tagStart = pos;
+    size_t tagLength = 0;
+    while (*pos && *pos != '|' && *(pos++) != ':') // we only advance pointer to one past if the last thing is : (not for |), this is intentional!
+        ++tagLength;
+    // ok, link data, skip to next |
+    char const* dataStart = pos;
+    size_t dataLength = 0;
+    while (*pos && *(pos++) != '|')
+        ++dataLength;
+    // ok, next should be link data end tag...
+    if (*(pos++) != 'h')
+        return nullptr;
+    // then visible link text, starts with [
+    if (*(pos++) != '[')
+        return nullptr;
+    // skip until we hit the next ], abort on unexpected |
+    char const* textStart = pos;
+    size_t textLength = 0;
+    while (*pos)
+    {
+        if (*pos == '|')
+            return nullptr;
+        if (*(pos++) == ']')
+            break;
+        ++textLength;
+    }
+    // link end tag
+    if (*(pos++) != '|' || *(pos++) != 'h' || *(pos++) != '|' || *(pos++) != 'r')
+        return nullptr;
+    // ok, valid hyperlink, return info
+    return { pos, color, tagStart, tagLength, dataStart, dataLength, textStart, textLength };
+}
+
+template <typename T>
+struct LinkValidator
+{
+    static bool IsTextValid(typename T::value_type, char const*, size_t) { return true; }
+    static bool IsColorValid(typename T::value_type, HyperlinkColor) { return true; }
+};
+
+// str1 is null-terminated, str2 is length-terminated, check if they are exactly equal
+static bool equal_with_len(char const* str1, char const* str2, size_t len)
+{
+    if (!*str1)
+        return false;
+    while (len && *str1 && *(str1++) == *(str2++))
+        --len;
+    return !len && !*str1;
+}
+
+template <>
+struct LinkValidator<LinkTags::achievement>
+{
+    static bool IsTextValid(AchievementLinkData const& data, char const* pos, size_t len)
+    {
+        if (!len)
+            return false;
+        for (uint8 i = 0; i < TOTAL_LOCALES; ++i)
+            if (equal_with_len(data.achievement->Title[i], pos, len))
+                return true;
+        return false;
+    }
+
+    static bool IsColorValid(AchievementLinkData const&, HyperlinkColor c)
+    {
+        return c == CHAT_LINK_COLOR_ACHIEVEMENT;
+    }
+};
+
+template <>
+struct LinkValidator<LinkTags::enchant>
+{
+    static bool IsTextValid(SpellInfo const* info, char const* pos, size_t len)
+    {
+        SkillLineAbilityMapBounds bounds = sSpellMgr->GetSkillLineAbilityMapBounds(info->Id);
```

Comment:
enchant links without related skill are allowed (client has code to construct them)
in which case the text is formatted as simply spellname, without "skillname: " prefix

Reply:
are you sure those aren't `spell:` links?